### ISSUE TITLE
PSNR differences between two different settings.

### ISSUE THREAD

**Issue by junhyukk** [xyq7/InvCompress#5]
Hi! Thanks for your great work!

During measuring PSNR values of test images, is there any difference between [them](https://github.com/xyq7/InvCompress/blob/main/codes/compressai/utils/eval_model/__main__.py#L176-L179)?
Since [the quantization method](https://github.com/xyq7/InvCompress/blob/main/codes/compressai/models/ours.py#L40) of 'forward()' is slightly different from [that](https://github.com/xyq7/InvCompress/blob/main/codes/compressai/models/priors.py#L557) of 'compress()', 
I'm wondering how much the difference occurs.

**Comment by felixcheng97** [comment]
Sorry for the late reply. We were busy for the new deadline before. Here are our replies.

**Question:** The difference between the inference_entropy_estimation() and inference().

https://github.com/xyq7/InvCompress/blob/17deb0f0e2f28432e6d5b38d0898ead58ccbcde4/codes/compressai/utils/eval_model/__main__.py#L176-L179

**Answer:** In short words, the 'forward()' function defined in network class is generally used for training, 

https://github.com/xyq7/InvCompress/blob/17deb0f0e2f28432e6d5b38d0898ead58ccbcde4/codes/compressai/models/ours.py#L34

while the 'compress()' and 'decompress()' functions defined in the network class are used for testing. 

https://github.com/xyq7/InvCompress/blob/17deb0f0e2f28432e6d5b38d0898ead58ccbcde4/codes/compressai/models/ours.py#L64

https://github.com/xyq7/InvCompress/blob/17deb0f0e2f28432e6d5b38d0898ead58ccbcde4/codes/compressai/models/ours.py#L102

The main difference between them is that 'forward()' function will not convert the latent features into actual bitstreams, while 'compress()' and 'decompress()' functions will. 

That is why you can see two modes in the evaluation script, which are implemented by the CompressAI library. One is the 'inference_entropy_estimation()' mode (shown below) using the 'forward()' function defined in the network class for entropy estimation of bits, which is somehow inaccurate (not much). 

https://github.com/xyq7/InvCompress/blob/17deb0f0e2f28432e6d5b38d0898ead58ccbcde4/codes/compressai/utils/eval_model/__main__.py#L133

Hence, all the testings in our paper are done using the 'inference()' mode (shown below), which makes use of the 'compress()' and 'decompress()' functions defined in the network class for accurate bits estimation.

https://github.com/xyq7/InvCompress/blob/17deb0f0e2f28432e6d5b38d0898ead58ccbcde4/codes/compressai/utils/eval_model/__main__.py#L81

Hope these help.

**Comment by xuezhongcailian** [comment]
Hi, Thank for your work.
I have a issue,if I If I have modified the JointAutoregressiveHierarchicalPriors model, it is changed to the compression and decompression of inference ，
![捕获](https://user-images.githubusercontent.com/31470730/156494482-ceb80ac9-50bc-4ae4-b03c-e443d0a63676.PNG)
![image](https://user-images.githubusercontent.com/31470730/156494628-9556c8d8-2d81-42a4-abaf-9165c5a25a5d.png)

This is different from the results derived from the source code. According to the theoretical calculation codedc process, there is no loss, is there any loss according to the original code decoding?

**Comment by felixcheng97** [comment]
Hi, thank you for your question. To be honest, I am not quite sure about your question. If I am understanding you correctly, here are my replies:

During inference (test time), you cannot use "noise" mode to get your y_hat. The "noise" is only used for training, since it will add random uniform (-0.5, 0.5) noise to y. What we need to do is interger rounding during test time, only by doing so no noise is introduced.

Hope this helps.

**Comment by xuezhongcailian** [comment]
![image](https://user-images.githubusercontent.com/31470730/157453883-e796aa52-ec76-4302-ae16-3bf0cde9c82b.png)
Hi, author，thanks for your work.
red line that I don't understand this code，is  the output  normalized? compute loss and psnr don't understand

**Comment by felixcheng97** [comment]
Thanks for your question.

For loss, we compute it between the raw output (without clamping) and gt.

For psnr and ssim, the raw output is first clamped to be in range (0, 1), which is processed in the torch2img function.

**Comment by xuezhongcailian** [comment]
but gt is (0,1), but psnr  max_val is not 1?

>

**Comment by felixcheng97** [comment]
please double check the definition of psnr, which can be defined with either max_val = 1 or 255.

**Comment by xuezhongcailian** [comment]
gt value is (0,1),&nbsp;raw output (without clamping), comput loss not understand.&nbsp;
At this time，&nbsp;comput psnr max_val is or not 1? Thank you











------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "xyq7/InvCompress"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年3月9日(星期三) 晚上10:56
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [xyq7/InvCompress] PSNR differences between two different settings. (#5)





 
please double check the definition of psnr, which can be defined with either max_val = 1 or 255.
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by felixcheng97** [comment]
Common practice is to compute loss between raw output and gt during training, so that we can have accurate and full gradients for backpropagation (this is because torch.clamp is non-differentiable outside the min-max range). 

During testing, we compute the loss between imgs for reference.
